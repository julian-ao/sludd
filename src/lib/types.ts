export type LocationQueryData = {
    metadata?: {
        totaltAntallTreff: number;
    };
    navn: LocationData[];
};

export type LocationData = {
    representasjonspunkt: {
        nord: number;
        øst: number;
    };
    stedsnavn: {
        skrivemåte: string;
    }[];
    stedsnummer: string;
    navneobjekttype?: string;
    kommuner?: {
        kommunenummer?: string;
        kommunenavn?: string;
    }[];
};

export type WeatherQueryData = {
    properties: {
        meta: {
            updated_at: string;
        };
        timeseries: {
            time: string;
            data: {
                instant: {
                    details: {
                        air_temperature: number;
                        wind_speed: number;
                        relative_humidity: number;
                        wind_from_direction: number;
                    };
                };
                next_1_hours: {
                    summary: {
                        symbol_code: string;
                    };
                    details: {
                        precipitation_amount: number;
                    };
                };
            };
        }[];
    };
};
