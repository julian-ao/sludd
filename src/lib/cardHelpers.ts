// Dictionary of weather descriptions for the different weather types
const weatherDescription: { [key: string]: string } = {
    clearsky: "Clear sky",
    fair: "Fair",
    partlycloudy: "Partly cloudy",
    cloudy: "Cloudy",
    lightrainshowers: "Light rain showers",
    rainshowers: "Rain showers",
    heavyrainshowers: "Heavy rain showers",
    lightrainshowersandthunder: "Light rain showers and thunder",
    rainshowersandthunder: "Rain showers and thunder",
    heavyrainshowersandthunder: "Heavy rain showers and thunder",
    lightsleetshowers: "Light sleet showers",
    sleetshowers: "Sleet showers",
    heavysleetshowers: "Heavy sleet showers",
    lightssleetshowersandthunder: "Light sleet showers and thunder",
    sleetshowersandthunder: "Sleet showers and thunder",
    heavysleetshowersandthunder: "Heavy sleet showers and thunder",
    lightsnowshowers: "Light snow showers",
    snowshowers: "Snow showers",
    heavysnowshowers: "Heavy snow showers",
    lightssnowshowersandthunder: "Light snow showers and thunder",
    snowshowersandthunder: "Snow showers and thunder",
    heavysnowshowersandthunder: "Heavy snow showers and thunder",
    lightrain: "Light rain",
    rain: "Rain",
    heavyrain: "Heavy rain",
    lightrainandthunder: "Light rain and thunder",
    rainandthunder: "Rain and thunder",
    heavyrainandthunder: "Heavy rain and thunder",
    lightsleet: "Light sleet",
    sleet: "Sleet",
    heavysleet: "Heavy sleet",
    lightsleetandthunder: "Light sleet and thunder",
    sleetandthunder: "Sleet and thunder",
    heavysleetandthunder: "Heavy sleet and thunder",
    lightsnow: "Light snow",
    snow: "Snow",
    heavysnow: "Heavy snow",
    lightsnowandthunder: "Light snow and thunder",
    snowandthunder: "Snow and thunder",
    heavysnowandthunder: "Heavy snow and thunder",
    fog: "Fog",
};

// Takes a symbol_code and returns the corresponding weather description
export function parseWeatherDescription(symbol_code: string) {
    const splitSymbolCode = symbol_code.split("_");
    const description = weatherDescription[splitSymbolCode[0]];

    return description;
}
// Dictionary of colors for the different weather types
const weatherColors = {
    clearsky: ["#ded364", "#ede26f", "#d9cf64"],
    fair: ["#8FEFBE", "#90F0AE", "#90F0CF"],
    cloudy: ["#8C9C9A", "#8D9C8E", "#9C998C"],
    rain: ["#8AD2EA", "#A4DCEE", "#93d1e6"],
    snow: ["#6E8FF5", "#6EA7F5", "#6EBEF5"],
    sleet: ["#9FA1ED", "#AB9FEC", "#B89FED"],
    fog: ["#5E6385", "#605E84", "#675E85"],
};

// Takes a symbol_code and returns the corresponding color
export function getColorFromWeatherDescription(symbol_code: string) {
    //use weatherColors to get color from symbol_code
    for (const [key, value] of Object.entries(weatherColors)) {
        if (symbol_code.includes(key)) {
            if (Array.isArray(value)) {
                return value[Math.floor(Math.random() * value.length)];
            }
            return value;
        }
    }
    return "#000000";
}
