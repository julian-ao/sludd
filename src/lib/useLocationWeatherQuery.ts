import { useEffect, useState } from "react";
import { useQuery } from "@tanstack/react-query";
import { LocationQueryData, WeatherQueryData } from "./types";

interface locationName {
    locationName: string;
    filter?: string;
}

interface locationId {
    locationId: string;
    filter?: string;
}

// Custom hook for fetching weather data based on locationName or locationId
export const useLocationWeatherQuery = (data: locationName | locationId) => {

    const [locationData, setLocationData] = useState<LocationQueryData | null>(
        null,
    );

    // Fetch location data (cordinates) based on locationName or locationId
    const locationQuery = useQuery<LocationQueryData, unknown>(
        ["location", data],
        async () => {
            if ("locationId" in data && data.locationId) {
                const url = `https://ws.geonorge.no/stedsnavn/v1/sted?stedsnummer=${
                    data.locationId
                }&utkoordsys=4258${
                    data.filter ? `&filtrer=${data.filter}` : ""
                }`;
                const res = await fetch(url);
                return res.json();
            } else if ("locationName" in data) {
                const base_url = `https://ws.geonorge.no/stedsnavn/v1/sted?sok=${data.locationName}&utkoordsys=4258&treffPerSide=1&side=1&filtrer=navn.representasjonspunkt,navn.stedsnavn.skrivemåte,navn.navneobjekttype,navn.stedsnummer`;
                const res = await fetch(base_url);
                return res.json();
            }

            throw new Error("Invalid data");
        },
    );

    // Set locationData when locationQuery is successful
    useEffect(() => {
        if (locationQuery.isSuccess) {
            setLocationData(locationQuery.data!);
        }
    }, [locationQuery]);

    // Fetch weather data based on cords from locationData
    const weatherQuery = useQuery<WeatherQueryData, unknown>(
        ["weather", locationData?.navn[0].representasjonspunkt, location],
        async () => {
            const res = await fetch(
                `https://api.met.no/weatherapi/locationforecast/2.0/compact?lat=${locationData?.navn[0].representasjonspunkt.nord}&lon=${locationData?.navn[0].representasjonspunkt.øst}`,
            );
            return res.json();
        },
        {
            enabled: !!locationData,
        },
    );

    return { locationData, weatherQuery };
};
