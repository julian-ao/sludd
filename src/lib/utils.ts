// Convert date to readable format
export const convertDateToReadable = (date: string) => {
    const month = date.slice(5, 7);
    const day = parseInt(date.slice(8, 10), 10).toString(); // Remove leading zero
    const hour = date.slice(11, 13);
    const minute = date.slice(14, 16);

    return `${day}. ${months[
        Number(month) - 1
    ].toLocaleLowerCase()} ${hour}:${minute}`;
};

const months = [
    "Januar",
    "Februar",
    "Mars",
    "April",
    "Mai",
    "Juni",
    "Juli",
    "August",
    "September",
    "Oktober",
    "November",
    "Desember",
];
