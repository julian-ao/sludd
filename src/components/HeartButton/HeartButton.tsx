import { useState } from 'react';
import './heartbutton.css';
import Popup from '../atoms/popup/Popup';
import solidHeart from '../../assets/heart-solid.svg';
import borderHeart from '../../assets/heart-border.svg';
import { LocationData } from '../../lib/types';

export type HeartButtonProps = {
    location: LocationData
};

const HeartButton = (props: HeartButtonProps) => {
    const [favorited, setFavorited] = useState<boolean>(() => {
        const favorites = JSON.parse(localStorage.getItem("favorites") || "[]");
        return favorites.map((f: LocationData) => f.stedsnummer).includes(props.location.stedsnummer);
    });

    const [isPopupShown, setIsPopupShown] = useState(false);

    const [popupText, setPopupText] = useState<string>();

    const toggleFavorite = () => {
        const favorites = JSON.parse(localStorage.getItem("favorites") || "[]");

        if (favorited) {
            const updatedFavorites = favorites.filter((f: LocationData) => f.stedsnummer !== props.location.stedsnummer);
            localStorage.setItem("favorites", JSON.stringify(updatedFavorites));
            setPopupText("ðŸ’” " + props.location.stedsnavn[0].skrivemÃ¥te + " fjernet fra favoritter");
        } else {
            favorites.push(props.location);
            localStorage.setItem("favorites", JSON.stringify(favorites));
            setPopupText("ðŸ’– " + props.location.stedsnavn[0].skrivemÃ¥te + " lagt til i favoritter");
        }

        setFavorited(!favorited);
        showPopup();
    };

    const showPopup = () => {
        setIsPopupShown(true);

        setTimeout(() => {
            setIsPopupShown(false);
        }, 1500);
    };

    return (
        <div>
            <div className='heart_div' onClick={toggleFavorite}>
                <img
                    className='solid'
                    style={favorited ? { opacity: 1 } : {}}
                    src={solidHeart}
                    alt='Hjerte'
                />
                <img src={borderHeart} alt='Hjerte' />
            </div>
            <Popup text={popupText || ''} show={isPopupShown} />
        </div>
    );
};

export default HeartButton;
