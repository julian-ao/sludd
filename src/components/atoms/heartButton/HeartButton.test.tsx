import { describe, vi } from "vitest";
import HeartButton from "./HeartButton";
import { render } from "@testing-library/react";
import { LocationData } from "../../../lib/types";

describe("HeartButton", () => {
    it("should render correctly", () => {
        const { container } = render(
            <HeartButton location={{
                representasjonspunkt: {
                    nord: 0,
                    øst: 0
                },
                stedsnavn: [],
                stedsnummer: "",
                navneobjekttype: undefined,
                kommuner: undefined
            }} />
        );
        expect(container).toMatchSnapshot();
    });

    it("should render correctly with favorite from localstorage", () => {
        const location = {
            representasjonspunkt: {
                nord: 0,
                øst: 0
            },
            stedsnavn: [
                { skrivemåte: "Test" }
            ],
            stedsnummer: "",
            navneobjekttype: undefined,
            kommuner: undefined
        } as LocationData;
        const localStorageMock = {
            getItem: vi.fn(),
            setItem: vi.fn(),
            clear: vi.fn()
        };
        Object.defineProperty(window, "localStorage", { value: localStorageMock });

        localStorageMock.getItem.mockReturnValueOnce(JSON.stringify([location]));

        const { container } = render(<HeartButton location={location} />);
        expect(container).toMatchSnapshot();
    });


    // it("should toggle favorite correctly", () => {
    //     const location = {
    //         representasjonspunkt: {
    //             nord: 0,
    //             øst: 0
    //         },
    //         stedsnavn: [{ skrivemåte: "Test" }],
    //         stedsnummer: "1",
    //         navneobjekttype: undefined,
    //         kommuner: undefined
    //     };
    //     const { getByTestId } = render(<HeartButton location={location} />);
    //     const heartButton = getByTestId("heart-button");

    //     fireEvent.click(heartButton);
    //     expect(localStorage.getItem("favorites")).toEqual(JSON.stringify([location]));
    //     expect(heartButton.getAttribute("class")).toContain("solid");

    //     fireEvent.click(heartButton);
    //     expect(localStorage.getItem("favorites")).toEqual(JSON.stringify([]));
    //     expect(heartButton.getAttribute("class")).not.toContain("solid");
    // });
});