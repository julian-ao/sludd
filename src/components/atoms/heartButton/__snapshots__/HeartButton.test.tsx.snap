// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeartButton > should render correctly 1`] = `
<div>
  <div
    class="heart-button"
  >
    <div
      class="heart_div"
    >
      <img
        alt="Hjerte"
        class="solid"
        src="/project1/src/assets/heart-solid.svg"
      />
      <img
        alt="Hjerte"
        src="/project1/src/assets/heart-border.svg"
      />
    </div>
    <div
      class="popup "
    >
      <div
        class="popup-content"
      >
        <p />
      </div>
    </div>
  </div>
</div>
`;

exports[`HeartButton > should render correctly with favorite from localstorage 1`] = `
<div>
  <div
    class="heart-button"
  >
    <div
      class="heart_div"
    >
      <img
        alt="Hjerte"
        class="solid"
        src="/project1/src/assets/heart-solid.svg"
        style="opacity: 1;"
      />
      <img
        alt="Hjerte"
        src="/project1/src/assets/heart-border.svg"
      />
    </div>
    <div
      class="popup "
    >
      <div
        class="popup-content"
      >
        <p />
      </div>
    </div>
  </div>
</div>
`;
